src/bucket/BucketManagerImpl.cpp

```diff
@@ -867,4 +868,165 @@ BucketManagerImpl::isShutdown() const
 {
     return mIsShutdown;
 }
+
+static void
+scanBucket(std::string const& pass, size_t sz, BucketInputIterator& in,
+           UnorderedMap<LedgerKey, LedgerEntry>* map = nullptr)
+{
+    medida::Timer timer;
+    timer.Time([&]() {
+        while (in)
+        {
+            if (map)
+            {
+                BucketEntry const& e = *in;
+                if (e.type() == LIVEENTRY || e.type() == INITENTRY)
+                {
+                    (*map)[LedgerEntryKey(e.liveEntry())] = e.liveEntry();
+                }
+                else
+                {
+                    if (e.type() != DEADENTRY)
+                    {
+                        throw std::runtime_error("Malformed bucket: unexpected "
+                                                 "non-INIT/LIVE/DEAD entry.");
+                    }
+                    map->erase(e.deadEntry());
```

Comment:
Should we add a sanity check here to make sure the key exists in the map? (to detect a malformed bucket right away)

Reply:
Fair. Fixed.